rust/worker/src/cache/cache.rs

```diff
@@ -0,0 +1,190 @@
+use crate::cache::config::CacheConfig;
+use crate::config::Configurable;
+use crate::errors::ChromaError;
+use async_trait::async_trait;
+use core::hash::Hash;
+use foyer::Cache as FoyerCache;
+use foyer::CacheBuilder;
+use foyer::LfuConfig;
+use foyer::LruConfig;
+use parking_lot::RwLock;
+use std::collections::HashMap;
+use std::sync::Arc;
+use thiserror::Error;
+
+#[derive(Clone)]
+pub(crate) enum Cache<K, V>
+where
+    K: Send + Sync + Hash + Eq + 'static,
+    V: Send + Sync + Clone + 'static,
+{
+    Unbounded(UnboundedCache<K, V>),
+    Foyer(FoyerCache<K, V>),
+}
+
+impl<K: Send + Sync + Hash + Eq + 'static, V: Send + Sync + Clone + 'static> Cache<K, V> {
+    pub fn new(config: &CacheConfig) -> Self {
+        match config {
+            CacheConfig::Unbounded(_) => Cache::Unbounded(UnboundedCache::new(config)),
+            _ => Cache::Foyer(FoyerCacheWrapper::new(config).cache),
+        }
+    }
+
+    pub fn insert(&self, key: K, value: V) {
+        match self {
+            Cache::Unbounded(cache) => cache.insert(key, value),
+            Cache::Foyer(cache) => {
+                cache.insert(key, value);
+            }
+        }
+    }
+
+    pub fn get(&self, key: &K) -> Option<V> {
+        match self {
+            Cache::Unbounded(cache) => cache.get(key),
+            Cache::Foyer(cache) => {
+                let entry = cache.get(key);
+                match entry {
+                    Some(v) => {
+                        let value = v.value().to_owned();
```

Comment:
does to_owned imply a clone? Should caches return references?

Reply:
Yes. This is an implicit clone. We are optimized for arc-typed values. I have considered alternatives such as returning references using unsafe Rust. However, after several iterations, I think returning value with clone is the safest approach given that recordbach is an arc type.